controllers/vspheredeploymentzone_controller_domain.go

```diff
@@ -0,0 +1,233 @@
+		default:
+			conditions.MarkTrue(ctx.VSphereDeploymentZone, infrav1.VSphereFailureDomainConfigurationCondition)
+		}
+	}
+	return nil
+}
+
+func (r vsphereDeploymentZoneReconciler) reconcileComputeCluster(ctx *context.VSphereDeploymentZoneContext) error {
+	if computeCluster := ctx.VSphereFailureDomain.Spec.Topology.ComputeCluster; computeCluster != nil {
+		ccr, err := ctx.AuthSession.Finder.ClusterComputeResource(ctx, *computeCluster)
+		if err != nil {
+			ctx.Logger.V(4).Error(err, "unable to find resource pool", "name", ctx.VSphereDeploymentZone.Spec.PlacementConstraint.ResourcePool)
+			return errors.Wrap(err, "unable to find resource pool")
+		}
+
+		if resourcePool := ctx.VSphereDeploymentZone.Spec.PlacementConstraint.ResourcePool; resourcePool != "" {
+			rp, err := ctx.AuthSession.Finder.ResourcePool(ctx, ctx.VSphereDeploymentZone.Spec.PlacementConstraint.ResourcePool)
+			if err != nil {
+				ctx.Logger.V(4).Error(err, "unable to find resource pool", "name", ctx.VSphereDeploymentZone.Spec.PlacementConstraint.ResourcePool)
+				return errors.Wrapf(err, "unable to find resource pool")
+			}
+
+			ref, err := rp.Owner(ctx)
+			if err != nil {
+				ctx.Logger.V(4).Error(err, "unable to find owner compute resource", "name", ctx.VSphereDeploymentZone.Spec.PlacementConstraint.ResourcePool)
+				return errors.Wrap(err, "unable to find resource pool")
+			}
+
+			if ref.Reference() != ccr.Reference() {
+				ctx.Logger.V(4).Error(nil, "compute cluster does not own resource pool", "computeCluster", computeCluster, "resourcePool", ctx.VSphereDeploymentZone.Spec.PlacementConstraint.ResourcePool)
+				return errors.Errorf("compute cluster %s does not own resource pool %s", *computeCluster, resourcePool)
+			}
+		}
+	}
+	return nil
+}
+
+// verifyFailureDomain verifies the Failure Domain. It verifies the existence of tag and category specified and
+// checks whether the specified tags exist on the DataCenter or Compute Cluster or Hosts (in a HostGroup).
+func (r vsphereDeploymentZoneReconciler) verifyFailureDomain(ctx *context.VSphereDeploymentZoneContext, failureDomain infrav1.FailureDomain) error {
+	if _, err := ctx.AuthSession.TagManager.GetTagForCategory(ctx, failureDomain.Name, failureDomain.TagCategory); err != nil {
+		return errors.Wrapf(err, "failed to verify tag %s and category %s", failureDomain.Name, failureDomain.TagCategory)
+	}
+
+	managedRefFunc := getManagedRefFinder(failureDomain, ctx.VSphereFailureDomain.Spec.Topology, ctx.AuthSession.Finder)
+	managedRefs, err := managedRefFunc(ctx)
+	if err != nil {
+		return err
+	}
+
+TagVerifier:
+	for _, ref := range managedRefs {
+		attachedTags, err := ctx.AuthSession.TagManager.GetAttachedTags(ctx, ref)
+		if err != nil {
+			return err
+		}
+		for _, tag := range attachedTags {
+			if tag.Name == failureDomain.Name {
+				continue TagVerifier
```

Comment:
why aren't we just breaking when we find the tag that matches ?

Reply:
The implicit assumption that I made here is that all the hosts in a host group should be tagged and hence I continue to loop through the rest of the references.
I thought about this but defensively assumed that we would want all hosts in the host group to be tagged. From the user's perspective, if you are using an explicit host group in the failure domain, all the hosts in that group are eligible for VM placement of VMs. And hence, it made sense for all of them to have the tag. 

But if that assumption is too rigid, we could assume that the validity of the failure domain depends on having at least one tagged host in the group. But that causes us to lose redundancy since all of the VM placement might happen on a single host.

Reply:
Relaxing the constraints to allow tags to be present on at least of the hosts in the host group. wdyt?

Reply:
hmmm no you're right, let's check all the hosts.